### ISSUE TITLE
No metrics on prometheus

### ISSUE THREAD

**Issue by karan-kapoor90** [oracle/docker-images#770]
@mriccell So I'm at the step where I've deployed and started the weblogic servers and the managed servers on the cluster and have also run the command to create the prometheus pod. I've got a prometheus pod running 

```
[opc@kube-master-01 ~]$ kubectl get pods
NAME                         READY     STATUS    RESTARTS   AGE
ms-0                         1/1       Running   0          1d
ms-1                         1/1       Running   0          1d
prometheus-dfcb484b5-q2brd   1/1       Running   0          1d
wls-admin-server-0           2/2       Running   0          1d
```

but dropdown menu is empty:

![image](https://user-images.githubusercontent.com/7921433/36616918-90b743e8-190b-11e8-82f3-a089ff17bafc.png)

So I went in and looked at the prometheus logs ot see if I could find anything there :

```
$ kubectl logs prometheus-dfcb484b5-q2brd
time="2018-02-22T16:34:02Z" level=error msg="github.com/prometheus/prometheus/discovery/kubernetes/kubernetes.go:207: Failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:default:default" cannot list pods at the cluster scope" component="kube_client_runtime" source="kubernetes.go:75"
time="2018-02-22T16:34:03Z" level=error msg="github.com/prometheus/prometheus/discovery/kubernetes/kubernetes.go:207: Failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:default:default" cannot list pods at the cluster scope" component="kube_client_runtime" source="kubernetes.go:75"
time="2018-02-22T16:34:04Z" level=error msg="github.com/prometheus/prometheus/discovery/kubernetes/kubernetes.go:207: Failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:default:default" cannot list p^C
```
So i went ahead and checked the prometheus configuration to find that it uses a prometheus.yml file (~~and I couldn't locate it anywhere in the git repo, assuming that it should have some wls specific config in it and hence it should be in the repo~~) :

```
$ kubectl exec prometheus-dfcb484b5-q2brd --stdin --tty -c prometheus /bin/sh

/prometheus # cat /etc/prometheus/prometheus.yml
global:
  scrape_interval:     5s
  external_labels:
    monitor: 'my-monitor'
scrape_configs:
- job_name: 'kubernetes-pods'
  kubernetes_sd_configs:
  - role: pod
  relabel_configs:
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
    action: keep
    regex: true
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
    action: replace
    target_label: __metrics_path__
    regex: (.+)
  - source_labels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
    action: replace
    regex: ([^:]+)(?::\d+)?;(\d+)
    replacement: $1:$2
    target_label: __address__
  - action: labelmap
    regex: __meta_kubernetes_pod_label_(.+)
  - source_labels: [__meta_kubernetes_pod_name]
    action: replace
    target_label: pod_name
  - regex: '(controller_revision_hash|job)'
    action: labeldrop
  - source_labels: [name]
    regex: '.*/(.*)$'
    replacement: $1
    target_label: webapp
  basic_auth:
   username: weblogic
   password: weblogic1/prometheus #
```

and the prometheus app (navigating to Status--> Configuration) shows the config as : 
```
global:
  scrape_interval: 5s
  scrape_timeout: 5s
  evaluation_interval: 1m
  external_labels:
    monitor: my-monitor
scrape_configs:
- job_name: kubernetes-pods
  scrape_interval: 5s
  scrape_timeout: 5s
  metrics_path: /metrics
  scheme: http
  kubernetes_sd_configs:
  - api_server: null
    role: pod
    namespaces:
      names: []
  basic_auth:
    username: weblogic
    password: <secret>
  relabel_configs:
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_scrape]
    separator: ;
    regex: "true"
    replacement: $1
    action: keep
  - source_labels: [__meta_kubernetes_pod_annotation_prometheus_io_path]
    separator: ;
    regex: (.+)
    target_label: __metrics_path__
    replacement: $1
    action: replace
  - source_labels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
    separator: ;
    regex: ([^:]+)(?::\d+)?;(\d+)
    target_label: __address__
    replacement: $1:$2
    action: replace
  - source_labels: []
    separator: ;
    regex: __meta_kubernetes_pod_label_(.+)
    replacement: $1
    action: labelmap
  - source_labels: [__meta_kubernetes_pod_name]
    separator: ;
    regex: (.*)
    target_label: pod_name
    replacement: $1
    action: replace
  - source_labels: []
    separator: ;
    regex: (controller_revision_hash|job)
    replacement: $1
    action: labeldrop
  - source_labels: [name]
    separator: ;
    regex: .*/(.*)$
    target_label: webapp
    replacement: $1
    action: replace
```

Not sure why prometheus is not showing the metrics. Any idea what I could be doing wrong?

**Comment by karan-kapoor90** [comment]
I went a step further and found that I needed to define a serviceAccountName in the deployment as mentioned in the final comment here https://github.com/prometheus/prometheus/issues/2763

In our case, there is no serviceAccount by the name prometheus so even though I tried it, the pod wouldn't even show up. Went back to using the default serviceAccount and atleast the pod shows up and the prometheus UI loads.

**Comment by karan-kapoor90** [comment]
Also noticed that one has to change the password in the promethus-kubernetes.yml file because as per the guide in the beginning, one is free to choose whatever password they like during building the image.. and in my case, I didn't choose the password to weblogic1 :)

**Comment by marinakog** [comment]
@karan-kapoor90 could you please verify the setting for basic_auth ? it should be valid user/password for WLS, somehow  in your configuration it shows 
basic_auth:
   username: weblogic
   password: weblogic1/prometheus #771 , in our demo we use 
basic_auth:
          username: weblogic
          password: welcome1

**Comment by mriccell** [comment]
@karan-kapoor90 did you resolve your problem with Prometheus?

**Comment by karan-kapoor90** [comment]
Sorry, have been travelling for work since the past couple of days, however coincidentally the OCI instances I set the entire thing up are having some kind of an issue because I have to setup the whole thing all over again, so as it stands today, I'm 10 steps back again and have to start from scratch. Please leave this open for a couple of days more while I try and get the whole thing working again.

**Comment by karan-kapoor90** [comment]
Ok, I'm back up.. any tips on how I can resolve the problem with prometheus?
- I've changed the password to match my password in the config
- the prometheus version is governed by the yaml file so no chance that's incorrect I guess.

**Comment by karan-kapoor90** [comment]
PS : The welcome1/prometheus # your see at the end of the file when I go into the pod is just the prompt getting printed at the end of file. Thats not a part of the file..

**Comment by marinakog** [comment]
Hi @karan-kapoor90 just to make sure we are using same config yaml, it is possible for you to : 
check the in the prometheus running targets ( go to config -target). Could it be possible that wls-exporter.war file is not deployed or started properly? Go to admin console of WLS cluster and make sure the application was deployed and started

**Comment by karan-kapoor90** [comment]
Checked, 
- wls-exporter is deployed to weblogic cluster
- Targets in the prometheus config is empty.. its an empty page..

prometheus logs still showing something that looks like a permission issue : 
```
time="2018-03-04T19:45:07Z" level=error msg="github.com/prometheus/prometheus/discovery/kubernetes/kubernetes.go:207: Failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:default:default" cannot list pods at the cluster scope" component="kube_client_runtime" source="kubernetes.go:75"
time="2018-03-04T19:45:08Z" level=error msg="github.com/prometheus/prometheus/discovery/kubernetes/kubernetes.go:207: Failed to list *v1.Pod: pods is forbidden: User "system:serviceaccount:default:default" cannot list pods at the cluster scope" component="kube_client_runtime" source="kubernetes.go:75"
```

**Comment by karan-kapoor90** [comment]
Random thought, but could the wls-exporter.war be bound to the domain name / location or some domain specific parameter? Because in the process of creating the oow-demo-domain, the "sample" path where it was getting me to put it at seemed to be giving some issue or the other, so I went with creating a default base_domain and then deploying the apps etc on that directly.

**Comment by marinakog** [comment]
@karan-kapoor90  One last thing: 1. create different name space, for example monitoring, use this link as example but you can delete alert part https://github.com/bhabermaas/kubernetes-projects/blob/master/kubernetes/prometheus-deployment.yaml , delete and create prometheus and grafana in monitoring name space, it is possible that there is conflict for default serviceaccount in system namespace. This is also most current configuration for prometheus. Could you please try it?

**Comment by karan-kapoor90** [comment]
@marinakog that sounds like a solution! I’d been investigating by looking back into the logs and it does looks like some sort of a service account issue. Is this something you just figured out or has it been addressed as a bug fix already and maybe I don’t have it?

Nevertheless, most definitely going to give this a shot the soonest I can.

**Comment by marinakog** [comment]
@karan-kapoor90 i just have tried prometheus.yaml provided in the demo with latest env and operator and could not reproduce your issue, so my guess , you just need to move prometheus pod to unique namespace instead of default to avoid conflict in your env. Please let me know if you solved it.

**Comment by karan-kapoor90** [comment]
Thank you @marinakog and @mriccell for all the help. I managed to get it working with a few changes. From the link @marinakog suggested, I picked up the configs to create a prometheus cluster role and cluster role binding, adding in a tiny bit to create a namespace called monitoring and it ran like a charm.

Prometheus was then showing all the weblogic metrics properly and I was able to feed it into a graph. Will submit a PR with the changes to the prometheus yaml.
